vernac/classes.ml

```diff
@@ -314,15 +314,10 @@ let instance_hook info global imps ?hook cst =
   (match hook with Some h -> h cst | None -> ())
 
 let declare_instance_constant info global imps ?hook name decl poly sigma term termtype =
-  (* XXX: Duplication of the declare_constant path *)
-  let sigma =
-    let levels = Univ.LSet.union (CVars.universes_of_constr termtype)
-                                 (CVars.universes_of_constr term) in
-    Evd.restrict_universe_context sigma levels
-  in
-  let uctx = Evd.check_univ_decl ~poly sigma decl in
   let kind = Decls.(IsDefinition Instance) in
-  let entry = Declare.definition_entry ~types:termtype ~univs:uctx term in
+  (* XXX: Type needs to be free of evars, not the body tho *)
+  let sigma, entry = DeclareDef.prepare_definition
+      ~allow_evars:true ~poly sigma decl ~types:(Some termtype) ~body:term in
   let kn = Declare.declare_constant ~name ~kind (Declare.DefinitionEntry entry) in
```

Comment:
Won't declare_constant die if the body has evars?

Reply:
Good question, when I wrote this I definitively needed the `allow_evars:true`, let me recall the details.